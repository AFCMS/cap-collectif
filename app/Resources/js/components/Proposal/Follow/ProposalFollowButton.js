// @flow
import React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import {
  Dropdown,
  MenuItem,
  Panel,
  ListGroup,
  ListGroupItem,
  FormGroup,
  Radio,
  Button,
  Popover,
  OverlayTrigger,
} from 'react-bootstrap';
import { FormattedMessage } from 'react-intl';
import FollowProposalMutation from '../../../mutations/FollowProposalMutation';
import UpdateFollowProposalMutation from '../../../mutations/UpdateFollowProposalMutation';
import type { ProposalFollowButton_proposal } from './__generated__/ProposalFollowButton_proposal.graphql';
import UnfollowProposalMutation from '../../../mutations/UnfollowProposalMutation';
import LoginOverlay from '../../Utils/LoginOverlay';

type Props = {
  proposal: ProposalFollowButton_proposal,
};

type State = {
  isJustFollowed: boolean,
};

export class ProposalFollowButton extends React.Component<Props, State> {
  state = {
    isJustFollowed: false,
  };
  componentWillReceiveProps(nextProps: Props) {
    if (this.props !== nextProps) {
      this.setState({
        isJustFollowed: !!nextProps.proposal.viewerIsFollowing,
      });
    }
  }
  changeFollowType(proposal: ProposalFollowButton_proposal, type: string) {
    if (
      proposal.viewerIsFollowing &&
      proposal.followerConfiguration !== null &&
      typeof proposal.followerConfiguration !== 'undefined'
    ) {
      return UpdateFollowProposalMutation.commit({
        input: {
          proposalId: proposal.id,
          notifiedOf: type,
        },
      });
    }
  }

  render() {
    const { proposal } = this.props;
    const { isJustFollowed } = this.state;
    if (!proposal.viewerIsFollowing) {
      return (
        <LoginOverlay>
          <Button
            className="btn btn-default proposal__button__follow"
            onClick={() => {
              return FollowProposalMutation.commit({
                input: { proposalId: proposal.id, notifiedOf: 'DEFAULT' },
              }).then(() => {
                this.setState({
                  isJustFollowed: true,
                });
                return true;
              });
            }}
            id={`proposal-follow-btn-${proposal.id}`}>
            <FormattedMessage id="follow" />
          </Button>
        </LoginOverlay>
      );
    }
    if (
      proposal.followerConfiguration !== null &&
      typeof proposal.followerConfiguration !== 'undefined'
    ) {
      return (
        <LoginOverlay>
          <span className="mb-0 proposal-follow-dropdown">
            <Dropdown
              className="mb-0 width250 fifty-shade-of-grey"
              id={`proposal-follow-btn-${proposal.id}`}
              defaultOpen={isJustFollowed}>
              <Dropdown.Toggle>
                <FormattedMessage id="following" />
              </Dropdown.Toggle>
              <Dropdown.Menu>
                <Panel
                  className="mb-0 bn width250"
                  header={
                    <span>
                      <FormattedMessage id="to-be-notified-of-new-of" />
                      <OverlayTrigger
                        placement="top"
                        overlay={
                          <Popover placement="top" className="in" id="pinned-label">
                            <FormattedMessage id="you-will-receive-a-summary-of-your-notifications-once-a-day" />
                          </Popover>
                        }>
                        <a>
                          <span className="cap-information" />
                        </a>
                      </OverlayTrigger>
                    </span>
                  }>
                  <form>
                    <FormGroup className="bn mb-0" id={`proposal-follow-btn-${proposal.id}`}>
                      <ListGroup className="mb-0">
                        <ListGroupItem className="">
                          <Radio
                            id={`proposal-follow-btn-default-${proposal.id}`}
                            name="default"
                            className="proposal__follow__advancement"
                            checked={
                              proposal.followerConfiguration.notifiedOf === 'DEFAULT'
                                ? 'checked'
                                : ''
                            }
                            inline
                            onClick={() => {
                              return this.changeFollowType(proposal, 'DEFAULT');
                            }}>
                            <b>
                              <FormattedMessage id="the-progress" />
                            </b>{' '}
                            <br />
                            <FormattedMessage id="list-of-progress-notifications" />
                          </Radio>
                        </ListGroupItem>
                        <ListGroupItem className="">
                          <Radio
                            name="default_and_comments"
                            id={`proposal-follow-btn-default_and_comments-${proposal.id}`}
                            className="proposal__follow__default_and_comments"
                            checked={
                              proposal.followerConfiguration.notifiedOf === 'DEFAULT_AND_COMMENTS'
                                ? 'checked'
                                : ''
                            }
                            onClick={() => {
                              return this.changeFollowType(proposal, 'DEFAULT_AND_COMMENTS');
                            }}>
                            <b>
                              <FormattedMessage id="progress-and-comments" />
                            </b>
                          </Radio>
                        </ListGroupItem>
                        <ListGroupItem className="">
                          <Radio
                            name="all"
                            id={`proposal-follow-btn-all-${proposal.id}`}
                            className="proposal__follow__all"
                            checked={
                              proposal.followerConfiguration.notifiedOf === 'ALL' ? 'checked' : ''
                            }
                            onClick={() => {
                              return this.changeFollowType(proposal, 'ALL');
                            }}>
                            <b>
                              <FormattedMessage id="all-activities" />
                            </b>
                            <br />
                            <FormattedMessage id="list-of-activity-notifications" />
                          </Radio>
                        </ListGroupItem>
                      </ListGroup>
                    </FormGroup>
                  </form>
                </Panel>
                <MenuItem
                  eventKey="1"
                  className="proposal__unfollow"
                  id={`proposal-unfollow-btn-${proposal.id}`}
                  onClick={() => {
                    if (proposal.viewerIsFollowing) {
                      return UnfollowProposalMutation.commit({
                        input: { proposalId: proposal.id },
                      }).then(() => {
                        this.setState({
                          isJustFollowed: false,
                        });
                      });
                    }
                  }}>
                  <FormattedMessage id="unfollow" />
                </MenuItem>
              </Dropdown.Menu>
            </Dropdown>
          </span>
        </LoginOverlay>
      );
    }
  }
}

export default createFragmentContainer(
  ProposalFollowButton,
  graphql`
    fragment ProposalFollowButton_proposal on Proposal {
      id
      viewerIsFollowing
      followerConfiguration {
        notifiedOf
      }
    }
  `,
);
