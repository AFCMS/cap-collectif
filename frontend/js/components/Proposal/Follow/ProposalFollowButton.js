// @flow
import React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
// TODO https://github.com/cap-collectif/platform/issues/7774
// eslint-disable-next-line no-restricted-imports
import {
  Dropdown,
  MenuItem,
  Panel,
  ListGroup,
  ListGroupItem,
  Radio,
  Button,
  OverlayTrigger,
} from 'react-bootstrap';
import { FormattedMessage } from 'react-intl';
import FollowProposalMutation from '../../../mutations/FollowProposalMutation';
import UpdateFollowProposalMutation from '../../../mutations/UpdateFollowProposalMutation';
import type { ProposalFollowButton_proposal } from '~relay/ProposalFollowButton_proposal.graphql';
import UnfollowProposalMutation from '../../../mutations/UnfollowProposalMutation';
import LoginOverlay from '../../Utils/LoginOverlay';
import Popover from '../../Utils/Popover';

import type { SubscriptionTypeValue } from '~relay/UpdateFollowProposalMutation.graphql';

type Props = {|
  +proposal: ?ProposalFollowButton_proposal,
  isAuthenticated: boolean,
|};

type State = {
  isJustFollowed: boolean,
};

export class ProposalFollowButton extends React.Component<Props, State> {
  // We could move this state to relay client store
  state = {
    isJustFollowed: false,
  };

  componentWillReceiveProps(nextProps: Props) {
    if (this.props !== nextProps) {
      this.setState({
        isJustFollowed: !!nextProps.proposal?.viewerIsFollowing,
      });
    }
  }

  changeFollowType(proposal: ProposalFollowButton_proposal, type: SubscriptionTypeValue) {
    if (
      proposal.viewerIsFollowing &&
      proposal.viewerFollowingConfiguration !== null &&
      typeof proposal.viewerFollowingConfiguration !== 'undefined'
    ) {
      return UpdateFollowProposalMutation.commit({
        input: {
          proposalId: proposal.id,
          notifiedOf: type,
        },
      });
    }
  }

  render() {
    const { proposal } = this.props;
    const { isJustFollowed } = this.state;
    if (!proposal)
      return (
        <Button
          disabled
          className="btn btn-default proposal__button__follow"
          id="proposal-follow-btn-placeholder">
          <i className="cap cap-rss mr-5" />
          <FormattedMessage id="follow" />
        </Button>
      );

    if (!proposal.viewerIsFollowing) {
      return (
        <LoginOverlay>
          <Button
            className="btn btn-default proposal__button__follow"
            onClick={() =>
              FollowProposalMutation.commit({
                input: { proposalId: proposal.id, notifiedOf: 'MINIMAL' },
              }).then(() => {
                this.setState({
                  isJustFollowed: true,
                });
                return true;
              })
            }
            id={`proposal-follow-btn-${proposal.id}`}>
            <i className="cap cap-rss mr-5" />
            <FormattedMessage id="follow" />
          </Button>
        </LoginOverlay>
      );
    }
    if (
      proposal.viewerFollowingConfiguration !== null &&
      typeof proposal.viewerFollowingConfiguration !== 'undefined'
    ) {
      return (
        <LoginOverlay>
          <span className="mb-0 custom-dropdown">
            <Dropdown
              className="mb-0 width250 custom-dropdown-bgd dropdown-button"
              id={`proposal-follow-btn-${proposal.id}`}
              defaultOpen={isJustFollowed}>
              <Dropdown.Toggle className="custom-dropdown-button">
                <FormattedMessage id="following" />
              </Dropdown.Toggle>
              <Dropdown.Menu>
                <Panel className="mb-0 b-none width250">
                  <Panel.Heading>
                    <FormattedMessage id="to-be-notified-by-email" />
                    <OverlayTrigger
                      placement="top"
                      overlay={
                        <Popover placement="top" className="in" id="pinned-label">
                          <FormattedMessage id="you-will-receive-a-summary-of-your-notifications-once-a-day" />
                        </Popover>
                      }>
                      <span className="cap-information ml-30" />
                    </OverlayTrigger>
                  </Panel.Heading>
                  <Panel.Body>
                    <div className="b-none mb-0" id={`proposal-follow-btn-${proposal.id}`}>
                      <ListGroup className="mb-0">
                        <ListGroupItem>
                          <Radio
                            id={`proposal-follow-btn-minimal-${proposal.id}`}
                            name="minimal"
                            className="proposal__follow__minimal"
                            checked={
                              proposal.viewerFollowingConfiguration === 'MINIMAL' ? 'checked' : ''
                            }
                            inline
                            onChange={() => this.changeFollowType(proposal, 'MINIMAL')}>
                            <b>
                              <FormattedMessage id="essential" />
                            </b>{' '}
                            <br />
                            <FormattedMessage id="updates-and-news" />
                          </Radio>
                        </ListGroupItem>
                        <ListGroupItem>
                          <Radio
                            name="essential"
                            id={`proposal-follow-btn-essential-${proposal.id}`}
                            className="proposal__follow__essential"
                            checked={
                              proposal.viewerFollowingConfiguration === 'ESSENTIAL' ? 'checked' : ''
                            }
                            onChange={() => this.changeFollowType(proposal, 'ESSENTIAL')}>
                            <b>
                              <FormattedMessage id="intermediate" />
                            </b>
                            <br />
                            <FormattedMessage id="updates-news-and-new-contributions" />
                          </Radio>
                        </ListGroupItem>
                        <ListGroupItem>
                          <Radio
                            name="all"
                            id={`proposal-follow-btn-all-${proposal.id}`}
                            className="proposal__follow__all"
                            checked={
                              proposal.viewerFollowingConfiguration === 'ALL' ? 'checked' : ''
                            }
                            onChange={() => this.changeFollowType(proposal, 'ALL')}>
                            <b>
                              <FormattedMessage id="complete" />
                            </b>
                            <br />
                            <FormattedMessage id="updates-news-new-contributions-votes-and-subscriptions" />
                          </Radio>
                        </ListGroupItem>
                      </ListGroup>
                    </div>
                  </Panel.Body>
                </Panel>
                <MenuItem
                  eventKey="1"
                  className="proposal__unfollow"
                  id={`proposal-unfollow-btn-${proposal.id}`}
                  onClick={() => {
                    if (proposal.viewerIsFollowing) {
                      return UnfollowProposalMutation.commit({
                        input: { proposalId: proposal.id },
                      }).then(() => {
                        this.setState({
                          isJustFollowed: false,
                        });
                      });
                    }
                  }}>
                  <FormattedMessage id="unfollow" />
                </MenuItem>
              </Dropdown.Menu>
            </Dropdown>
          </span>
        </LoginOverlay>
      );
    }
    return null;
  }
}

export default createFragmentContainer(ProposalFollowButton, {
  proposal: graphql`
    fragment ProposalFollowButton_proposal on Proposal
      @argumentDefinitions(isAuthenticated: { type: "Boolean!" }) {
      id
      viewerIsFollowing @include(if: $isAuthenticated)
      viewerFollowingConfiguration @include(if: $isAuthenticated)
    }
  `,
});
