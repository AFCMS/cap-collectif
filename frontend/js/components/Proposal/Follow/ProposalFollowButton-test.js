// @flow
/* eslint-env jest */
import { graphql } from 'react-relay';
import ProposalFollowButton from './ProposalFollowButton';

describe('<ProposalFollowButton />', () => {
  const proposalViewIsFollowing = {
    id: 'proposal1',
    viewerIsFollowing: true,
    viewerFollowingConfiguration: 'MINIMAL',
  };
  const proposalViewIsNotFollowing = {
    id: 'proposal1',
    viewerIsFollowing: false,
    viewerFollowingConfiguration: null,
  };

  const query = graphql`
    query ProposalFollowButtonTestQuery @relay_test_operation {
      proposal: node(id: "test-id") {
        ...ProposalFollowButton_proposal
      }
    }
  `;

  it('should render a button to unfollow a proposal when viewer is following.', () => {
    expect(
      global.renderWithRelay(ProposalFollowButton, {
        query,
        spec: {
          Proposal() {
            return proposalViewIsFollowing;
          },
        },
      }),
    ).toMatchSnapshot();
  });

  it('should render a button to follow a proposal when viewer is not following.', () => {
    expect(
      global.renderWithRelay(ProposalFollowButton, {
        query,
        spec: {
          Proposal() {
            return proposalViewIsNotFollowing;
          },
        },
      }),
    ).toMatchSnapshot();
  });
});
