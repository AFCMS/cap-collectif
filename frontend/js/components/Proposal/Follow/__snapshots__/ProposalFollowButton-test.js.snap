// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProposalFollowButton /> should render a button to follow a proposal when viewer is not following. 1`] = `
<Connect(LoginOverlay)>
  <button
    className="btn btn-default proposal__button__follow"
    id="proposal-follow-btn-proposal1"
    onClick={[Function]}
    type="button"
  >
    <i
      className="cap cap-rss mr-5"
    />
    <FormattedMessage
      id="follow"
      values={Object {}}
    />
  </button>
</Connect(LoginOverlay)>
`;

exports[`<ProposalFollowButton /> should render a button to unfollow a proposal when viewer is following. 1`] = `
<Connect(LoginOverlay)>
  <span
    className="mb-0 custom-dropdown"
  >
    <Uncontrolled(Dropdown)
      className="mb-0 width250 custom-dropdown-bgd dropdown-button"
      defaultOpen={false}
      id="proposal-follow-btn-proposal1"
    >
      <DropdownToggle
        bsClass="dropdown-toggle"
        bsRole="toggle"
        className="custom-dropdown-button"
        open={false}
        useAnchor={false}
      >
        <FormattedMessage
          id="following"
          values={Object {}}
        />
      </DropdownToggle>
      <DropdownMenu
        bsClass="dropdown-menu"
        bsRole="menu"
        pullRight={false}
      >
        <Uncontrolled(Panel)
          className="mb-0 b-none width250"
        >
          <PanelHeading
            bsClass="panel"
            componentClass="div"
          >
            <FormattedMessage
              id="to-be-notified-by-email"
              values={Object {}}
            />
            <OverlayTrigger
              defaultOverlayShown={false}
              overlay={
                <CapcoPopover
                  className="in"
                  id="pinned-label"
                  placement="top"
                >
                  <FormattedMessage
                    id="you-will-receive-a-summary-of-your-notifications-once-a-day"
                    values={Object {}}
                  />
                </CapcoPopover>
              }
              placement="top"
              trigger={
                Array [
                  "hover",
                  "focus",
                ]
              }
            >
              <span
                className="cap-information ml-30"
              />
            </OverlayTrigger>
          </PanelHeading>
          <PanelBody
            bsClass="panel"
            collapsible={false}
          >
            <div
              className="b-none mb-0"
              id="proposal-follow-btn-proposal1"
            >
              <ListGroup
                bsClass="list-group"
                className="mb-0"
              >
                <ListGroupItem
                  bsClass="list-group-item"
                  listItem={false}
                >
                  <Radio
                    bsClass="radio"
                    checked="checked"
                    className="proposal__follow__minimal"
                    disabled={false}
                    id="proposal-follow-btn-minimal-proposal1"
                    inline={true}
                    name="minimal"
                    onChange={[Function]}
                    title=""
                  >
                    <b>
                      <FormattedMessage
                        id="essential"
                        values={Object {}}
                      />
                    </b>
                     
                    <br />
                    <FormattedMessage
                      id="updates-and-news"
                      values={Object {}}
                    />
                  </Radio>
                </ListGroupItem>
                <ListGroupItem
                  bsClass="list-group-item"
                  listItem={false}
                >
                  <Radio
                    bsClass="radio"
                    checked=""
                    className="proposal__follow__essential"
                    disabled={false}
                    id="proposal-follow-btn-essential-proposal1"
                    inline={false}
                    name="essential"
                    onChange={[Function]}
                    title=""
                  >
                    <b>
                      <FormattedMessage
                        id="intermediate"
                        values={Object {}}
                      />
                    </b>
                    <br />
                    <FormattedMessage
                      id="updates-news-and-new-contributions"
                      values={Object {}}
                    />
                  </Radio>
                </ListGroupItem>
                <ListGroupItem
                  bsClass="list-group-item"
                  listItem={false}
                >
                  <Radio
                    bsClass="radio"
                    checked=""
                    className="proposal__follow__all"
                    disabled={false}
                    id="proposal-follow-btn-all-proposal1"
                    inline={false}
                    name="all"
                    onChange={[Function]}
                    title=""
                  >
                    <b>
                      <FormattedMessage
                        id="complete"
                        values={Object {}}
                      />
                    </b>
                    <br />
                    <FormattedMessage
                      id="updates-news-new-contributions-votes-and-subscriptions"
                      values={Object {}}
                    />
                  </Radio>
                </ListGroupItem>
              </ListGroup>
            </div>
          </PanelBody>
        </Uncontrolled(Panel)>
        <MenuItem
          bsClass="dropdown"
          className="proposal__unfollow"
          disabled={false}
          divider={false}
          eventKey="1"
          header={false}
          id="proposal-unfollow-btn-proposal1"
          onClick={[Function]}
        >
          <FormattedMessage
            id="unfollow"
            values={Object {}}
          />
        </MenuItem>
      </DropdownMenu>
    </Uncontrolled(Dropdown)>
  </span>
</Connect(LoginOverlay)>
`;

exports[`<ProposalFollowButton /> should render correctly when not authenticated 1`] = `
<Button
  active={false}
  block={false}
  bsClass="btn"
  bsStyle="default"
  className="btn btn-default proposal__button__follow"
  disabled={true}
  id="proposal-follow-btn-placeholder"
>
  <i
    className="cap cap-rss mr-5"
  />
  <FormattedMessage
    id="follow"
    values={Object {}}
  />
</Button>
`;
